labelCloud/io/labels/kitti.py

```diff
@@ -20,72 +23,130 @@ def __init__(
-                    bbox.set_rotations(
-                        0, 0, rel2abs_rotation(-float(line_elements[14]) + math.pi / 2)
-                    )
-                else:
-                    bbox.set_rotations(0, 0, rel2abs_rotation(float(line_elements[14])))
-                bbox.set_classname(line_elements[0])
-                labels.append(bbox)
+                rotation = -float(meta["rotation_y"]) + math.pi / 2 if self.transformed else float(meta["rotation_y"])
+                bbox.set_rotations(0, 0, rel2abs_rotation(rotation))
+                bbox.set_classname(meta["type"])
+                bboxes.append(bbox)
             logging.info("Imported %s labels from %s." % (len(label_lines), label_path))
-        return labels
+        return bboxes
 
     def export_labels(self, bboxes: List[BBox], pcd_path: Path) -> None:
         data = str()
 
         # Labels
-        for bbox in bboxes:
+        for i, bbox in enumerate(bboxes):
             obj_type = bbox.get_classname()
             centroid = bbox.get_center()
-            if self.transformed:
-                centroid = (-centroid[1], centroid[2] + 2.3, centroid[0])
-            location = " ".join([str(self.round_dec(v)) for v in centroid])
             dimensions = bbox.get_dimensions()
             if self.transformed:
-                dimensions = (dimensions[2], dimensions[1], dimensions[0])
+                centroid = (
+                    centroid[0],
+                    centroid[1],
+                    centroid[2] - dimensions[2] / 2,
+                )  # centroid in KITTI located on bottom face of bbox
+                dimensions = dimensions[2], dimensions[1], dimensions[0]
+                xyz1 = np.insert(np.asarray(centroid), 3, values=[1])
+                xyz1 = self.T_v2c @ xyz1
+                centroid = tuple([float(n) for n in xyz1[:-1]])
+            location_str = " ".join([str(self.round_dec(v)) for v in centroid])
             dimensions_str = " ".join([str(self.round_dec(v)) for v in dimensions])
-            rotation_z = bbox.get_z_rotation()
-            if self.transformed:
-                rotation_y = self.round_dec(
-                    -(abs2rel_rotation(rotation_z) - math.pi / 2)
-                )
-            else:
-                rotation_y = self.round_dec(abs2rel_rotation(rotation_z))
-
-            data += (
-                " ".join(
-                    [
-                        obj_type,
-                        "0 0 0 0 0 0 0",
-                        dimensions_str,
-                        location,
-                        str(rotation_y),
-                    ]
-                )
-                + "\n"
-            )
+            rotation = bbox.get_z_rotation()
+            rotation = abs2rel_rotation(rotation)
+            rotation = -(rotation - math.pi / 2) if self.transformed else rotation
+            rotation = str(self.round_dec(rotation))
+
+            out_str = list(self.bboxes_meta[i].values())
```

Comment:
I get your intent now with the `meta` attribute. The thing is that the exact sequence of bounding boxes might change and as soon as you add a bounding box it crashes.

But I will try to adapt to a dictionary which uses the `id(label` as a key. This should allow us to keep the mapping.

- [ ] ensure correct mapping when readding label properties

In the future a metadata attribute on the label object itself might be a cleaner approach.